connection_params.ts

```diff
@@ -78,20 +81,26 @@ export class ConnectionParams {
       config = dsn;
     }
 
-    this.database = selectFrom([config, pgEnv], "database");
-    this.host = selectFrom([config, pgEnv, DEFAULT_CONNECTION_PARAMS], "host");
-    this.port = selectFrom([config, pgEnv, DEFAULT_CONNECTION_PARAMS], "port");
-    this.user = selectFrom([config, pgEnv], "user");
+    let potentiallyNull: { [K in keyof IConnectionParams]?: string } = {
+      database: selectFrom([config, pgEnv], "database"),
+      user: selectFrom([config, pgEnv], "user")
+    };
+
+    this.host =
+      selectFrom([config, pgEnv], "host") || DEFAULT_CONNECTION_PARAMS["host"];
+    this.port =
+      selectFrom([config, pgEnv], "port") || DEFAULT_CONNECTION_PARAMS["port"];
+    this.application_name =
+      selectFrom([config, pgEnv], "application_name") ||
+      DEFAULT_CONNECTION_PARAMS["application_name"];
     this.password = selectFrom([config, pgEnv], "password");
-    this.application_name = selectFrom(
-      [config, pgEnv, DEFAULT_CONNECTION_PARAMS],
-      "application_name"
-    );
 
     const missingParams: string[] = [];
 
-    ["database", "user"].forEach(param => {
-      if (!this[param]) {
+    (["database", "user"] as Array<keyof IConnectionParams>).forEach(param => {
```

Comment:
I tried to rework this as a ```for key in situation```, but typescript kept trying to downgrade the type of ```key``` to ```string```, and then then not let me set it on ```this``` because it was a string.

I was also trying to keep the structure as close to what you had, and hold on to the behavior of getting all the errors rather than just the one. Not sure if you have any other ideas, but i did spend a while on this one :/

Reply:
Okay, I'll dig into this case in the evening, let's leave as it is now.